sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -764,35 +894,119 @@ public Write withoutValidation() {
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setAttemptTimeout(timeoutMs).build())
+          .build();
     }
 
     /**
-     * Returns a new {@link BigtableIO.Write} that will use an official Bigtable emulator.
+     * Returns a new {@link BigtableIO.Write} with the operation timeout in milliseconds for writes.
      *
-     * <p>This is used for testing.
+     * <p>Does not modify this object.
+     */
+    public Write withOperationTimeout(long timeoutMs) {
+      checkArgument(timeoutMs > 0, "operation timeout must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setOperationTimeout(timeoutMs).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with the retry delay in milliseconds.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withRetryInitialDelay(long delayMs) {
+      checkArgument(delayMs > 0, "delay must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setRetryInitialDelay(delayMs).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with retry multiplier.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withRetryDelayMultiplier(double multiplier) {
+      checkArgument(multiplier > 0, "multiplier must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setRetryDelayMultiplier(multiplier).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with the number of elements in a batch.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withBatchElements(long size) {
+      checkArgument(size > 0, "batch element size must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setBatchElements(size).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with the number of bytes in a batch.
+     *
+     * <p>Does not modify this object.
+     */
+    public Write withBatchBytes(long size) {
+      checkArgument(size > 0, "batch byte size must be positive");
+      BigtableWriteOptions options = getBigtableWriteOptions();
+      return toBuilder()
+          .setBigtableWriteOptions(options.toBuilder().setBatchBytes(size).build())
+          .build();
+    }
+
+    /**
+     * Returns a new {@link BigtableIO.Write} with the max number of concurrent requests.
+     *
+     * <p>Does not modify this object.
      */
+    public Write withMaxRequests(long requests) {
```

Comment:
Why did we name this maxRequests? Shouldnt we align with veneer and do max outstanding elements?

Reply:
I think it's easier to do the calculation than setting outstandingElements and outstandingBytes, which might confuse people..

Reply:
Im not sure why this would confuse people. The calculation is a rough estimate and not quite correct (ie when there are partial batches and retries). If we had previously exposed maxRequests on the public surface then I can see this calculation making sense. But since this is green field, why not stay consistent with veneer?